ports/esp32/esp32_pcnt.c

```diff
@@ -0,0 +1,456 @@
+/*
+ * This file is part of the MicroPython project, http://micropython.org/
+ *
+ * The MIT License (MIT)
+ *
+ * Copyright (c) 2021-22 Jonathan Hogg
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+
+#include "py/runtime.h"
+#include "py/mphal.h"
+#include "py/obj.h"
+
+#if MICROPY_PY_ESP32_PCNT
+
+#include "modesp32.h"
+#include "driver/pcnt.h"
+
+typedef struct _esp32_pcnt_obj_t {
+    mp_obj_base_t base;
+    pcnt_unit_t unit;
+    struct _esp32_pcnt_obj_t *next;
+    mp_obj_t handler;
+    // Set in the IRQ handler to indicate that this unit is ready.
+    uint32_t pending;
+} esp32_pcnt_obj_t;
+
+// Linked list of PCNT units.
+MP_REGISTER_ROOT_POINTER(struct _esp32_pcnt_obj_t *esp32_pcnt_obj_head);
+
+// Once off installation of the PCNT ISR service (using the default service).
+// Persists across soft reset.
+STATIC bool pcnt_isr_service_installed = false;
+
+// Reserve a dedicated static scheduler node to avoid needing to contend for a
+// slot (it's important not to miss events leading to skipped counts).
+STATIC mp_sched_node_t pcnt_irq_sched_node = {0};
+
+// Scheduler callback. At least one unit has raised an IRQ, walk the list and
+// invoke the Python handlers.
+STATIC void IRAM_ATTR esp32_pcnt_run_irq_handlers(mp_sched_node_t *node) {
+    (void)node;
+    esp32_pcnt_obj_t *pcnt = MP_STATE_PORT(esp32_pcnt_obj_head);
+    while (pcnt != NULL) {
+        uint32_t status = pcnt->pending;
+        pcnt->pending ^= status; // TODO should this disable IRQs?
```

Comment:
Does this TODO need to be resolved, or can the comment be removed?

Reply:
I think I've convinced myself the TODO is resolved.